### DISCUSSION TITLE
Server Side Rendering performance remark

### DISCUSSION THREAD

**Discussion by wmuizelaar** [vercel/next.js#67495] [Help]
### Summary

**Summary**

We are having issues getting NextJS to run our mostly SSR application with good performance. We’re not sure if there is anything we can do to improve, or if there is anything we are doing wrong, so would like feedback on this matter.

**Our situation**

We are migrating a large site towards NextJS. It is currently on Next version 14.1.0, using the Pages Router.  We will mostly be Serverside rendering, due to SEO constraints. The application gets all of its information from one GraphQL API. We have little we can cache, due to high degree of personalization. So far one large page has been migrated, but we’re running into performance issues with this one page, both latency wise and resource wise.

We run in Kubernetes, and during our initial tests we could not get above 4.5 requests per second per pod without having our 90p latency balloon. We only use half a CPU when we arrive at this 90p threshold (of 500ms, not unreasonably fast), and increasing the amount of traffic per pod makes the latency go up skyhigh.

**Further observations**
We have intensely investigated our code paths to try and optimize them and cannot find any glaring issues. Furthermore, we have been testing different GC settings, achieving minor reductions in event loop lag. However, from enabling tracing we have gotten some great insights.

The application mostly seems to have two parts: Get ServerSideProps, which contains of multiple GraphQL calls being done, and AppRender.renderToReadableStream, which is the actual rendering of the page.

As we understand it, the rendering of an app essentially blocks the entire event loop of the running application, until it is done. For us, this rendering step usually takes around 70ms (from the duration of the span in our tracing), though sometimes dropping to 40ms or spiking up to 100ms.

What we see is that much of our additional latency comes from gaps between GraphQL calls being finished by the upstream API (as we can see in the spans from our outbound Istio proxy) and NextJS finishing the fetch. As an example, NextJS has a span for an API call that takes 90ms, but our Istio outbound proxy span only takes 20ms.

Without fail, each and every single one of these gaps can be explained by a render step from a different request on the same pod. Sometimes the handling of the API call can even have two interruptions from different renders, and then the latency really starts ramping up. This is slightly amplified by the fact that we have to perform sequential calls to our (federated) GraphQL API due to data ordering.

This seems to suggest that the optimal way to run this on Kubernetes is with tiny pods, each having such little load that the chance of suffering from another request’s render step is very small. That's not very ideal.

**_We have the following questions:_**

1. For large server-side rendered pages, is this kind of performance normal or to be expected? Did we go in with wrong expectations, or have we simply made some kind of mistake somewhere? Is it possible to do the rendering off the event-loop, so we are not blocking this anymore?

2. For large server-side rendered pages, is this rendering time (around 70ms) to be expected, or wildly above the norm?
3. Do you have any other suggestions for what we can look at?

**Comment by m-sanders** [comment]
@wmuizelaar did you get anywhere further with this? I am dealing with a similar setup and seeing similar performance issues with server-side rendering.

**Reply by wmuizelaar** [reply]
Unfortunately not. While trying to make it faster, by reducing the amount of data, dom-nodes, etcetera, and that all works, but it's not the silver bullet we're looking for. It would be so ideal if you could easily fork off the rendering to worker threads, but we haven't found a solution with NextJS for that so far.

**Comment by cupofjoakim** [comment]
We're observing the same thing in our setup. There are some differences in the details (our latency does not balloon until ~1.1 CPU, we run a ton of pods, etc). From what I gather, this seems to be the norm when relying heavily on SSR in nextjs, ideally it seems that you want to do SSG when possible.

That being said, last week we found some "free" performance by removing the use of `next/image`. We ran tons of images, and since our image providers have image optimisations in their apis, the image component didn't actually do anything, except for hogging up a lot of CPU. Removing it gave us ~3 extra requests per second, but then again, we ran a ton of images.

When it comes to general performance, we've resorted to trying to work around the issue, by offloading as much work as possible to our page cache on a third party cdn and we plan to make our pages as cacheable as possible. As in, we serve a cached version to all users and then enrich the view clientside for personalized content. Does not solve the original low performance, but atleast alleviates it a bit.

**Comment by anderslauri** [comment]
Hej,

@cupofjoakim @wmuizelaar Thanks for this thread. Coming from the outside - did you manage to come up with further improvements to increase throughput? Background. We are also using SSR to render pages in an e-commerce setup. In our case CPU is low, however, ELU is spiking relatively fast (we use ELU to scale out in GKE). We have estimated we can serve around 6,5 req/s per pod given this setup - which is horrendously low (my opinion).

Our workaround for the moment is offloading most of the work to Varnish (north of nextjs) to ensure we can apply a micro-caching strategy when serving dynamic content. This works well - around 80% of traffic is served from cache.

**Comment by wmuizelaar** [comment]
What we've done eventually is migrate away from NextJS. Caching would not work for us unfortunately because of the highly personalized content that we're serving. With Remix as alternative for our React-SSR-pages we are able to achieve twice as much req/s/pod.

**Comment by cupofjoakim** [comment]
@anderslauri From our end we're still in the same situation. We've been able to gain a bit more performance, so we're now hovering around 12-15 req/s/pod, but it's been very hard to get here and it's still nowhere near good enough IMO.

The main issue has been getting the event loop lag under control, and since we do huge DOM sizes serialisation has been the big thief. There also still seems to be something going on when a pod has many connections running and it's doing some long fetch request to our underlying services - it's as if it can't have as many active connections as it should. This has been confirmed by running a backend that just responded instantly with mocked data - all of a sudden the event loop lag kind of disappears.

All in all we are at least now at a place where we don't feel like we need to make a swap to a different framework, but we're definitely still investigating possible performance gains.
